frontend/model/contracts/group.js

```diff
@@ -441,28 +568,51 @@ DefineContract({
           }
         }
       },
-      process ({ data, meta }, { state }) {
-        Vue.delete(state.profiles, data.member)
+      process ({ data, meta }, { state, getters }) {
+        state.profiles[data.member].status = PROFILE_STATUS.REMOVED
+        state.profiles[data.member].departedDate = meta.createdDate
       },
-      async sideEffect ({ data }) {
-        await sbp('gi.sideEffects/group/removeMember', {
-          username: data.member,
-          groupId: data.groupId
-        })
+      async sideEffect ({ data, contractID }, { state }) {
+        const rootState = sbp('state/vuex/state')
+        const contracts = rootState.contracts || {}
+
+        if (data.member === rootState.loggedIn.username) {
+          // If this member is re-joining the group, ignore the rest
+          // so the member doesn't remove themself again.
+          if (sbp('okTurtles.data/get', 'JOINING_GROUP')) {
+            return
+          }
+
+          const groupIdToSwitch = Object.keys(contracts)
+            .find(cID => contracts[cID].type === 'group' &&
+                  cID !== contractID &&
+                  rootState[cID].settings) || null
+
+          sbp('state/vuex/commit', 'setCurrentGroupId', groupIdToSwitch)
+          sbp('state/vuex/commit', 'removeContract', contractID)
+          sbp('controller/router').push({ path: groupIdToSwitch ? '/dashboard' : '/' })
+          // TODO - #828 remove other group members contracts if applicable
+        } else {
+          // TODO - #828 remove the member contract if applicable.
+          // sbp('state/vuex/commit', 'removeContract', getters.groupProfile(data.member).contractID)
+        }
+        // TODO - #850 verify open proposals and see if they need some re-adjustment.
       }
     },
     'gi.contracts/group/removeOurselves': {
-      validate: objectOf({
-        groupId: string
+      validate: objectMaybeOf({
+        reason: string
```

Comment:
Everywhere `removeOurselves` is called, I didn't see `reason` to be called in any one of the places. In what case should a reason be given? And how can the user add it in the UI?

Reply:
They might be able to give an explanation for why they're leaving. It's optional though so I don't think it's a big deal if we leave it here? Harder to add back after the protocol is fully defined.